### PR TITLE
afpi: Added support for `loginWithOtp` 

### PR DESCRIPTION
**Pull request by wferem1**
- [x] All new/changed/fixed functionality is covered by tests (or N/A)
- [x] I have added documentation for all new/changed functionality (or N/A)

### 📋 Changes

* Expanded the plugin with the `loginWithOtp` method, both for iOS and Android.
* Added a convenience method in `ApiException` to get the `mfaToken` from the `ApiException` when `mfaIsRequired` is true

If you receive an `ApiException` with `mfaIsRequired` set to true, you can now prompt the user for a OTP code and call `Auth0.api.loginWithOtp` with the `mfaToken` present in the `ApiException` and the user's otp.

### 📎 References
This PR solves part of the features requested in #168

### CODE REVIEW

File: auth0_flutter/android/src/main/kotlin/com/auth0/auth0_flutter/AuthenticationExceptionExtensions.kt [hunk 1]
```diff
@@ -3,32 +3,33 @@ package com.auth0.auth0_flutter
...
+            "isNetworkError" to ex.isNetworkError,
+            "isBrowserAppNotAvailable" to ex.isBrowserAppNotAvailable,
+            "isPKCENotAvailable" to ex.isPKCENotAvailable,
+            "isInvalidAuthorizeURL" to ex.isInvalidAuthorizeURL,
+            "isInvalidConfiguration" to ex.isInvalidConfiguration,
+            "isCanceled" to ex.isCanceled,
+            "isPasswordLeaked" to ex.isPasswordLeaked,
+            "isLoginRequired" to ex.isLoginRequired,
+        ))
+        if (ex.getValue("mfa_token") != null) { put("mfa_token", ex.getValue("mfa_token")!!) }
```
Comment (Widcket) [inline auth0_flutter/android/src/main/kotlin/com/auth0/auth0_flutter/AuthenticationExceptionExtensions.kt:31]: Let's use camelCase for the key.

File: auth0_flutter/android/src/main/kotlin/com/auth0/auth0_flutter/AuthenticationExceptionExtensions.kt [hunk 2]
```diff
@@ -3,31 +3,33 @@ package com.auth0.auth0_flutter
...
-            "isBrowserAppNotAvailable" to this.isBrowserAppNotAvailable,
-            "isPKCENotAvailable" to this.isPKCENotAvailable,
-            "isInvalidAuthorizeURL" to this.isInvalidAuthorizeURL,
-            "isInvalidConfiguration" to this.isInvalidConfiguration,
-            "isCanceled" to this.isCanceled,
-            "isPasswordLeaked" to this.isPasswordLeaked,
-            "isLoginRequired" to this.isLoginRequired,
-        )
-    )
+    var ex = this;
```
Comment (Widcket) [inline auth0_flutter/android/src/main/kotlin/com/auth0/auth0_flutter/AuthenticationExceptionExtensions.kt:6]: Let's use a non-abbreviated word for better readability, e.g. `exception`.

File: auth0_flutter_platform_interface/lib/src/method_channel_auth0_flutter_auth.dart [hunk 3]
```diff
@@ -15,6 +16,7 @@ import 'user_profile.dart';
 
 const MethodChannel _channel = MethodChannel('auth0.com/auth0_flutter/auth');
 const String authLoginMethod = 'auth#login';
+const String authLoginWithOtpMethod = 'auth#login_otp';
```
Comment (Widcket) [inline auth0_flutter_platform_interface/lib/src/method_channel_auth0_flutter_auth.dart:19]: All the other method keys use camelCase, please use it for this new key as well.

File: auth0_flutter/android/src/main/kotlin/com/auth0/auth0_flutter/request_handlers/api/LoginApiRequestHandler.kt [hunk 4]
```diff
@@ -44,6 +44,7 @@ class LoginApiRequestHandler : ApiRequestHandler {
         if (args.getOrDefault("parameters", null) is HashMap<*, *>) {
             loginBuilder.addParameters(args["parameters"] as Map<String, String>)
         }
+        loginBuilder.validateClaims();
```
Comment (Widcket) [inline auth0_flutter/android/src/main/kotlin/com/auth0/auth0_flutter/request_handlers/api/LoginApiRequestHandler.kt:47]: I don't think we should enable this without also providing a way to customize the ID Token validation (e.g. the issuer -> `withIdTokenVerificationIssuer ()` in Auth0.Android). Since this would be outside the scope of this PR, better leave this for another PR.

File: auth0_flutter/android/src/main/kotlin/com/auth0/auth0_flutter/request_handlers/api/LoginOtpApiRequestHandler.kt [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+        val args = request.data
+
+        assertHasProperties(listOf("mfaToken", "otp"), args)
+
+        val loginBuilder = api
+            .loginWithOTP(
+                args["mfaToken"] as String,
+                args["otp"] as String,
+            )
+        loginBuilder.validateClaims();
```
Comment (Widcket) [inline auth0_flutter/android/src/main/kotlin/com/auth0/auth0_flutter/request_handlers/api/LoginOtpApiRequestHandler.kt:35]: Same as above.

File: auth0_flutter/ios/Classes/AuthAPI/AuthAPILoginWithOtpMethodHandler.swift [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+        guard let otp = arguments[Argument.otp] as? String else {
+            return callback(FlutterError(from: .requiredArgumentMissing(Argument.otp.rawValue)))
+        }
+        guard let mfaToken = arguments[Argument.mfaToken] as? String else {
+            return callback(FlutterError(from: .requiredArgumentMissing(Argument.mfaToken.rawValue)))
+        }
+
+        client
+            .login(withOTP: otp,
+                  mfaToken: mfaToken)
```
Comment (Widcket) [inline auth0_flutter/ios/Classes/AuthAPI/AuthAPILoginWithOtpMethodHandler.swift:22]:
```suggestion
            .login(withOTP: otp, mfaToken: mfaToken)
```

File: auth0_flutter/lib/src/authentication_api.dart [hunk 7]
```diff
@@ -73,6 +73,31 @@ class AuthenticationApi {
         parameters: parameters,
       )));
 
+  /// Authenticates the user using a [mfaToken] and a [otp], after [login] returned with an [ApiException] with [ApiException.isMultifactorRequired] set to true.
```
Comment (Widcket) [inline auth0_flutter/lib/src/authentication_api.dart:76]:
```suggestion
  /// Authenticates the user using a [mfaToken] and an [otp], after [login] returned with an [ApiException] with [ApiException.isMultifactorRequired] set to true.
```

File: auth0_flutter/lib/src/authentication_api.dart [hunk 8]
```diff
@@ -73,6 +73,31 @@ class AuthenticationApi {
...
+  ///
+  /// ## Endpoint docs
+  /// https://auth0.com/docs/api/authentication#verify-with-one-time-password-otp-
+  ///
+  /// ## Usage example
+  ///
+  /// ```dart
+  /// final result = await auth0.api.loginWithOtp({
+  ///   otp: '123456',
+  ///   mfaToken: 'receivedMfaToken'
```
Comment (Widcket) [inline auth0_flutter/lib/src/authentication_api.dart:88]:
```suggestion
  ///   mfaToken: 'received_mfa_token'
```

Comment (Widcket) [reply auth0_flutter/lib/src/authentication_api.dart:88]: Let's continue the convention followed by other API docs from the same package.

File: auth0_flutter/android/src/main/kotlin/com/auth0/auth0_flutter/AuthenticationExceptionExtensions.kt [hunk 9]
```diff
@@ -3,31 +3,33 @@ package com.auth0.auth0_flutter
...
+            "isNetworkError" to ex.isNetworkError,
+            "isBrowserAppNotAvailable" to ex.isBrowserAppNotAvailable,
+            "isPKCENotAvailable" to ex.isPKCENotAvailable,
+            "isInvalidAuthorizeURL" to ex.isInvalidAuthorizeURL,
+            "isInvalidConfiguration" to ex.isInvalidConfiguration,
+            "isCanceled" to ex.isCanceled,
+            "isPasswordLeaked" to ex.isPasswordLeaked,
+            "isLoginRequired" to ex.isLoginRequired,
+        ))
+        if (ex.getValue("mfa_token") != null) { put("mfa_token", ex.getValue("mfa_token")!!) }
```
Comment (Widcket) [inline auth0_flutter/android/src/main/kotlin/com/auth0/auth0_flutter/AuthenticationExceptionExtensions.kt:31]: Let's use camelCase for the key.

File: auth0_flutter_platform_interface/lib/src/auth/api_exception.dart [hunk 10]
```diff
@@ -35,6 +35,10 @@ class ApiException extends Auth0Exception {
 
   bool get isMultifactorRequired =>
       _errorFlags.getBooleanOrFalse('isMultifactorRequired');
+  String? get mfaToken =>
+      details.containsKey('mfa_token') && details['mfa_token'] is String
```
Comment (Widcket) [inline auth0_flutter_platform_interface/lib/src/auth/api_exception.dart:39]: Same: let's use camelCase for the key.

File: auth0_flutter/android/src/main/kotlin/com/auth0/auth0_flutter/AuthenticationExceptionExtensions.kt (hunk 2 above)
Comment (wferem1) [reply auth0_flutter/android/src/main/kotlin/com/auth0/auth0_flutter/AuthenticationExceptionExtensions.kt:6]: I'm not the most experienced with kotlin, but the `buildMap` method was my solution to have an optional key-value pair in the map, instead of changing the signature of the map from <String, Any> to <String, Any?>. In this method referencing `this` would point to the map instead of the exception, which is why I had to introduce this variable.
If you have a better method, feel free to let me know

File: auth0_flutter/android/src/main/kotlin/com/auth0/auth0_flutter/AuthenticationExceptionExtensions.kt (hunk 1 above)
Comment (wferem1) [reply auth0_flutter/android/src/main/kotlin/com/auth0/auth0_flutter/AuthenticationExceptionExtensions.kt:31]: on iOS, this key is already present in `Auth0APIError.details` with key `mfa_token`, which is why i chose to do have the same key on the android side. I could replace this key in `auth0_flutter/ios/Classes/Extensions.swift:47`,

**Comment by Widcket** [comment]
> Since I made changes in the platform_interface, the plugin would need these changes, I could add a dependency override,

We typically make the dependency override when developing, but change it back before releasing. Please change it so we can see if the CI passes.

File: auth0_flutter/android/src/main/kotlin/com/auth0/auth0_flutter/AuthenticationExceptionExtensions.kt (hunk 2 above)
Comment (Widcket) [reply auth0_flutter/android/src/main/kotlin/com/auth0/auth0_flutter/AuthenticationExceptionExtensions.kt:6]: No worries, I'm no Kotlin expert myself either. I'll ask @poovamraj to review the Android bits once we're close to merging.

File: auth0_flutter/android/src/main/kotlin/com/auth0/auth0_flutter/AuthenticationExceptionExtensions.kt (hunk 1 above)
Comment (Widcket) [reply auth0_flutter/android/src/main/kotlin/com/auth0/auth0_flutter/AuthenticationExceptionExtensions.kt:31]: Do you mean here https://github.com/auth0/auth0-flutter/blob/main/auth0_flutter/ios/Classes/Extensions.swift#L47?

Comment (wferem1) [reply auth0_flutter/android/src/main/kotlin/com/auth0/auth0_flutter/AuthenticationExceptionExtensions.kt:31]: Yes, self.info` already contains the mfa token, but with `mfa_token` as key, I started with `mfaToken` until I tried testing the plugin on an iOS device.

File: auth0_flutter/android/src/main/kotlin/com/auth0/auth0_flutter/request_handlers/api/LoginApiRequestHandler.kt (hunk 4 above)
Comment (wferem1) [reply auth0_flutter/android/src/main/kotlin/com/auth0/auth0_flutter/request_handlers/api/LoginApiRequestHandler.kt]: I removed these

File: auth0_flutter/android/src/main/kotlin/com/auth0/auth0_flutter/AuthenticationExceptionExtensions.kt (hunk 1 above)
Comment (Widcket) [reply auth0_flutter/android/src/main/kotlin/com/auth0/auth0_flutter/AuthenticationExceptionExtensions.kt]: Yes, that's right.

**Comment by Widcket** [comment]
@poovamraj could you please review the Kotlin/Android bits?

**Comment by Widcket** [comment]
@wferem1 could you please correct the title? Once you edit it, the corresponding CI check will run again.

File: auth0_flutter/android/src/test/kotlin/com/auth0/auth0_flutter/request_handlers/api/LoginWithOtpApiRequestHandlerTest.kt [hunk 11]
```diff
@@ -0,0 +1,170 @@
...
+        assertThat(
+            exception.message,
+            equalTo("Required property 'otp' is not provided.")
+        )
+    }
+
+    @Test
+    fun `should throw when missing mfaToken`() {
+        val options =
+            hashMapOf("otp" to "test-otp")
```
Comment (poovamraj) [inline auth0_flutter/android/src/test/kotlin/com/auth0/auth0_flutter/request_handlers/api/LoginWithOtpApiRequestHandlerTest.kt:52]:
```suggestion
        val options = hashMapOf("otp" to "test-otp")
```

File: auth0_flutter/android/src/test/kotlin/com/auth0/auth0_flutter/request_handlers/api/LoginWithOtpApiRequestHandlerTest.kt [hunk 12]
```diff
@@ -0,0 +1,170 @@
...
+import org.robolectric.RobolectricTestRunner
+import java.text.SimpleDateFormat
+import java.util.*
+
+@RunWith(RobolectricTestRunner::class)
+class LoginWithOtpApiRequestHandlerTest {
+    @Test
+    fun `should throw when missing otp`() {
+        val options =
+            hashMapOf("mfaToken" to "test-mfa-token")
```
Comment (poovamraj) [inline auth0_flutter/android/src/test/kotlin/com/auth0/auth0_flutter/request_handlers/api/LoginWithOtpApiRequestHandlerTest.kt:28]:
```suggestion
        val options = hashMapOf("mfaToken" to "test-mfa-token")
```

**Review by poovamraj — COMMENTED** [review]
Made a few small suggestions. Should be easy.

**Comment by Widcket** [comment]
@wferem1 are you still interested in this?

**Comment by wferem1** [comment]
@Widcket 
Sorry, I was a bit busy, and because of security reasons, I decided to go back to the web auth login flow. 
The reason I wanted to support this, was that the browser on the device stores cookies, so that the OTP in the web auth flow is only asked once every 30 days.

However, allowing these login methods, seems less secure than the PKCE flow. 

I can try and take a look at the remarks in my free time, as it seems like a good addition for those who want to use the OTP method. But it's not an urgent request for me anymore.

File: auth0_flutter/lib/src/authentication_api.dart [hunk 13]
```diff
@@ -73,6 +73,31 @@ class AuthenticationApi {
         parameters: parameters,
       )));
 
+  /// Authenticates the user using a [mfaToken] and an [otp], after [login] returned with an [ApiException] with [ApiException.isMultifactorRequired] set to true.
```
Comment (Widcket) [inline auth0_flutter/lib/src/authentication_api.dart:76]:
```suggestion
  /// Authenticates the user using a [mfaToken] and an [otp], after [login] returned with an [ApiException] with [ApiException.isMultifactorRequired] set to `true`.
```

File: auth0_flutter_platform_interface/lib/src/auth0_flutter_auth_platform.dart [hunk 14]
```diff
@@ -30,6 +31,11 @@ abstract class Auth0FlutterAuthPlatform extends PlatformInterface {
     throw UnimplementedError('authLogin() has not been implemented');
   }
 
+  Future<Credentials> loginWithOtp(
+      final ApiRequest<AuthLoginWithOtpOptions> request) {
+    throw UnimplementedError('authLogin() has not been implemented');
```
Comment (Widcket) [inline auth0_flutter_platform_interface/lib/src/auth0_flutter_auth_platform.dart:36]:
```suggestion
    throw UnimplementedError('authLoginWithOtp() has not been implemented');
```

**Review by Widcket — APPROVED** [review]
